### PR TITLE
Upload-only link shares

### LINKED ISSUE #7747: Feature request: Write-only anonymous upload
Anonymous upload with write-only permissions would be handy for gathering data from multiple external people. This would make it unnecessary to create separate accounts for that purpose.

It would also be safer than the password under the file sharing options, in case the password feature still loads a MD5 hashed password to the user's browser cache. (This was found in audit in our company early 2013. Unfortunately I never got the details so don't ask.)

### PR DESCRIPTION
**Pull request by PVince81**
Fixes https://github.com/owncloud/core/issues/7747.

Checklist here: https://github.com/owncloud/core/issues/7747#issuecomment-282263457

There's more to come, we can't merge this as is because backend support is required.

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] BUG: how to solve conflicts ?

- [ ] TODO: JS unit tests for the new view
- [ ] TODO: beautify the view (cc @felixheidecke)

**Comment by PVince81** [comment]
This is using the file-upload.js code so when this https://github.com/owncloud/core/pull/26306#issuecomment-290762884 is fixed to work with chunking, the button from anonymous upload should also be able to work with chunking.

@DeepDiver1975 @tomneedham @IljaN @butonic check this out

**Comment by PVince81** [comment]
- [ ] TODO: error handling, properly display error message
   - [ ] conflict
   - [ ] folder gone / whatever
   - [ ] insufficient space

**Comment by PVince81** [comment]
@pmaier1 I think having it called "Allow file listing" and set by default is better.

This is how it looks on this PR:

files_drop_core

If you disable "Allow uploads" the option "Allow file listing" becomes greyed out and checked.

**Comment by pmaier1** [comment]
Ok nice so far. I don't like the wording "Allow viewing file listing" :) "Hide file listing" would be perfect IMO.
As you would need to change your logic probably call it "Show file listing".

**Comment by PVince81** [comment]
The part I dislike about "Hide file listing" is mostly because you need to think positive for "Allow uploads" and then negative "Hide" instead of "Show".

We could also rename "Allow uploads" to "Disallow uploads".

@felixheidecke any feedback on the wordings ?

**Comment by pmaier1** [comment]
I leave this up to you. 'Allow' and 'Show' or 'Disallow' and 'Hide' is both fine for me while I like the first pair a bit more :)

**Comment by PVince81** [comment]
- [ ] TODO: add capability entry

**Comment by PVince81** [comment]
- [ ] Change "Allow upload" to "Allow editing"

**Comment by PVince81** [comment]
- [ ] "Show file listing"

**Comment by PVince81** [comment]
Added integration tests.

One covers the file name deduplication and will fail currently as I still need to implement this.

**Comment by PVince81** [comment]
Capability added "supports_upload_only"

**Comment by PVince81** [comment]
Found a relatively clean way to deduplicate names by overriding the PUT method, but only when a header "X-OC-Autorename" is set.

@DeepDiver1975 can you have a look and tell me whether the approach is acceptable ? It's all in https://github.com/owncloud/core/pull/27548/commits/c47b09b2987a0fdde9a5dacea492cbb9a22b6b63

**Comment by PVince81** [comment]
As discussed with @DeepDiver1975 in a chat, the approach is acceptable.
One change:

- [ ] rename "X-OC-Autorename" to "OC-Autorename"

**Comment by PVince81** [comment]
Added JS unit tests.

Next up: beautify and proper error messages

**Comment by tomneedham** [comment]
@butonic

**Comment by PVince81** [comment]
Uh oh, seems the push of JS unit tests from the other computer didn't work. Note to self: no rebase!

**Comment by PVince81** [comment]
- [ ] cherry pick commits from my other computer/branch which weren't pushed properly (OC-Autorename + JS unit tests)

**Comment by PVince81** [comment]
- [ ] TEST: drag and drop upload

**Comment by PVince81** [comment]
- [ ] design adjustments:
    - [ ] visible dropzone with text "Select files for upload"
        - [ ] clicking dropzone opens file picker
        - [ ] dropping on dropzone uploads files directly

**Comment by PVince81** [comment]
files-drop

File: tests/integration/features/sharing-v1.feature [hunk 1]
```diff
@@ -263,6 +264,27 @@ Feature: sharing
...
 	And Downloaded content should be "test2"
 
+  Scenario: Uploading file to a public upload-only share that was deleted does not work
+    Given user "user0" exists
+    And As an "user0"
+    And creating a share with
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+    And User "user0" deletes file "/FOLDER"
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: This should be 'When'.

File: tests/integration/features/sharing-v1.feature [hunk 2]
```diff
@@ -263,6 +264,27 @@ Feature: sharing
...
+	Then publicly uploading a file does not work
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Too much indentation here.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/sharing-v1.feature [hunk 3]
```diff
@@ -216,6 +216,75 @@ Feature: sharing
...
+
+  Scenario: Uploading to a public upload-only share with password
+    Given user "user0" exists
+    And As an "user0"
+    And creating a share with
+      | path | FOLDER |
+      | shareType | 3 |
+      | password | publicpw |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with password "publicpw" and content "test"
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Indentation from here to the end of the scenario is not aligned.

File: tests/integration/features/sharing-v1.feature [hunk 4]
```diff
@@ -216,6 +216,75 @@ Feature: sharing
...
+    And the HTTP status code should be "404"
+
+  Scenario: Uploading same file to a public upload-only share multiple times
+    Given user "user0" exists
+    And As an "user0"
+    And creating a share with
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with content "test"
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Indentation from here to the end of the scenario is not aligned.

File: tests/integration/features/sharing-v1.feature [hunk 5]
```diff
@@ -216,6 +216,75 @@ Feature: sharing
...
       | mimetype | httpd/unix-directory |
 
+  Scenario: Uploading to a public upload-only share
+    Given user "user0" exists
+    And As an "user0"
+    And creating a share with
+      | path | FOLDER |
+      | shareType | 3 |
+      | permissions | 4 | 
+	And publicly uploading file "test.txt" with content "test"
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Indentation from here to the end of the scenario is not aligned.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/sharing-v1.feature [hunk 6]
```diff
@@ -1,1162 +1,1231 @@
...
+			| permissions | 19 |
+			| stime | A_NUMBER |
+			| storage | A_NUMBER |
+			| mail_send | 0 |
+			| uid_owner | user0 |
+			| storage_id | home::user0 |
+			| file_parent | A_NUMBER |
+			| share_with_displayname | user1 |
+			| displayname_owner | user0 |
+			| mimetype			| text/plain |
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Too separated.

File: tests/integration/features/sharing-v1.feature [hunk 7]
```diff
@@ -1,1162 +1,1231 @@
...
+			| file_target | /textfile0.txt |
+			| permissions | 1 |
+			| stime | A_NUMBER |
+			| storage | A_NUMBER |
+			| mail_send | 0 |
+			| uid_owner | user0 |
+			| storage_id | home::user0 |
+			| file_parent | A_NUMBER |
+			| displayname_owner | user0 |
+			| mimetype			| text/plain |
```
Comment (SergioBertolinSG) [inline tests/integration/features/sharing-v1.feature]: Too separated.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/sharing-v1.feature (hunk 6 above)
Comment (PVince81) [inline tests/integration/features/sharing-v1.feature]: I did a search and replace of two spaces "  " with tabs. Apparently for some reason there were two spaces there. I'll adjust it.

**Comment by PVince81** [comment]
Rebased and fixes review comments

**Comment by PVince81** [comment]
Going to pre-review the diff and fix missing stuff.

**Comment by PVince81** [comment]
Ok, fixed everything I found when looking at the diff.

Please review! See whitespace-less diff: https://github.com/owncloud/core/pull/27548/files?w=1

@pmaier1 @DeepDiver1975 @tomneedham @felixheidecke

File: apps/files_sharing/js/PublicUploadView.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+			if (!this._itemTemplate) {
+				this._itemTemplate = Handlebars.compile(ITEM_TEMPLATE);
+			}
+			return this._itemTemplate(data);
+		},
+
+		render: function () {
+			this.$el.html(this.template({
+				title: t('files_sharing', 'Anonymous upload'),
+				uploadButtonLabel: t('files_sharing', 'Select files for upload'),
```
Comment (felixheidecke) [inline apps/files_sharing/js/PublicUploadView.js]: I suggest *"Drop files here for upload"* to hint to the drag & drop ability.

Comment (pmaier1) [inline apps/files_sharing/js/PublicUploadView.js]: "Click here to select files or use drag & drop to upload" as you can also use the file picker I assume.

Comment (PVince81) [inline apps/files_sharing/js/PublicUploadView.js]: I'll wait for you guys to settle on a solution, let me know 😄

**Review by DeepDiver1975 — APPROVED** [review]
NICE

File: apps/files_sharing/js/PublicUploadView.js (hunk 8 above)
Comment (pmaier1) [inline apps/files_sharing/js/PublicUploadView.js]: My final call: "Click to select files or use drag & drop to upload"

**Comment by PVince81** [comment]
I'll go with @pmaier1's text for now and we can merge.

If it's not optimal we can change it in a subsequent PR. This will make it possible to already get this tested...

**Comment by PVince81** [comment]
Tests passed on previous commit and the text is not covered by unit tests => merge

**Comment by PVince81** [comment]
> Tests passed on previous commit and the text is not covered by unit tests => merge

Never trust the green ticks, Jenkins wasn't prevent, there's a failure on master now 💥

**Comment by butonic** [comment]
Is there a migration path when upgrading from an oc 9.x with files_drop app?

**Comment by PVince81** [comment]
@butonic no migration path possible for technical reasons (route name is different).

It was decided to say in the release notes that the shares must be recreated.

cc @pmaier1

**Comment by ckamm** [comment]
@PVince81 I'm adding the relevant functionality to the client. To clarify: Previously the client checked the "Allow editing" box if create and update permission were present. This seems to have changed? To get matching behavior I must only check for create permission.

Does the update permission still toggle anything ui wise?

**Pull request merged**